javatests/arcs/core/data/expression/TypeCheckerTest.kt

```diff
@@ -0,0 +1,300 @@
+      SeqType(TextType),
+      scope
+    )
+
+    checkAssignableTo(
+      PaxelParser.parse("from c in company from p in c.people select p.name"),
+      SeqType(TextType),
+      scope
+    )
+  }
+
+  @Test
+  fun check_where_expression() {
+    val emptyScope = MapScope<InferredType>("root", mapOf())
+    val scope = emptyScope.builder().apply {
+      set("numbers", SeqType(IntType))
+    }.build()
+
+    checkAssignableTo(
+      PaxelParser.parse("from p in numbers where p < 10 select p"),
+      SeqType(IntType),
+      scope
+    )
+  }
+
+  @Test
+  fun check_select_expression() {
+    val emptyScope = MapScope<InferredType>("root", mapOf())
+    val scope = emptyScope.builder().apply {
+      set("numbers", SeqType(IntType))
+    }.build()
+
+    checkAssignableTo(
+      PaxelParser.parse("from p in numbers select new Foo { x: p, y: p + 1i, z: p == 2 }"),
+      SeqType(
+        ScopeType(
+          emptyScope.builder("Foo").set("x", IntType).set("y", IntType).set("z", BooleanType)
+            .build()
+        )
+      ),
+      scope
+    )
+  }
+
+  @Test
+  fun check_let_expression() {
+    val emptyScope = MapScope<InferredType>("root", mapOf())
+    val scope = emptyScope.builder().apply {
+      set("numbers", SeqType(IntType))
+    }.build()
+
+    checkAssignableTo(
+      PaxelParser.parse("from p in numbers let x = (p > 5) select x"),
+      SeqType(BooleanType),
+      scope
+    )
+  }
+
+  @Test
+  fun check_orderby_expression() {
+    val emptyScope = MapScope<InferredType>("root", mapOf())
+    val scope = emptyScope.builder().apply {
+      set("numbers", SeqType(IntType))
+    }.build()
+
+    checkAssignableTo(
+      PaxelParser.parse("from p in numbers orderby p select p"),
+      SeqType(IntType),
+      scope
+    )
+  }
+
+  @Test
+  fun check_function_expression() {
+    val emptyScope = MapScope<InferredType>("root", mapOf())
+    val scope = emptyScope.builder().apply {
+      set("numbers", SeqType(IntType))
+      set("words", SeqType(TextType))
+    }.build()
+
+    checkAssignableTo(
+      PaxelParser.parse("from p in words select now()"),
+      SeqType(LongType),
+      scope
+    )
+
+    checkAssignableTo(
+      PaxelParser.parse("from p in numbers select average(p)"),
+      SeqType(DoubleType),
+      scope
+    )
+
+    checkAssignableTo(
+      PaxelParser.parse("from p in numbers select min(p)"),
+      SeqType(IntType),
+      scope
+    )
+
+    checkAssignableTo(
+      PaxelParser.parse("from p in numbers select max(p)"),
+      SeqType(IntType),
+      scope
+    )
+
+    checkAssignableTo(
+      PaxelParser.parse("from p in numbers select sum(p)"),
+      SeqType(IntType),
+      scope
+    )
+
+    checkAssignableTo(
+      PaxelParser.parse("from p in words select count(numbers)"),
+      SeqType(IntType),
+      scope
+    )
+
+    checkAssignableTo(
+      PaxelParser.parse("from p in numbers select first(words)"),
+      SeqType(UnionType(TextType, NullType)),
```

Comment:
Do you think representing nullable types as a Union(Type, Null) is a good idea? I'm genuinely not sure...

Reply:
Technically all types are sets, e.g.

boolean = {true, false}
ints = {0, 1, ..., inf}
text = char*

High types are just products fo these with set union or intersection, so there's nothing wrong with using union types for this. We could add a special wrapper like NullType(InferredType) to make a nullable version, but it would have forced a lot of extra checks elsewhere, whereas unions are general purpose. 

We can change it later if there's good reason.